src/MSBuildLocator/DotNetSdkLocationHelper.cs

```diff
@@ -169,39 +172,55 @@ private static IntPtr HostFxrResolver(Assembly assembly, string libraryName)
             {
-                if (key == NativeMethods.hostfxr_resolve_sdk2_result_key_t.resolved_sdk_dir)
+                int rc = NativeMethods.hostfxr_resolve_sdk2(exe_dir: dotnetPath, working_dir: workingDirectory, flags: 0, result: (key, value) =>
                 {
-                    resolvedSdk = value;
-                }
-            });
+                    if (key == NativeMethods.hostfxr_resolve_sdk2_result_key_t.resolved_sdk_dir)
+                    {
+                        resolvedSdk = value;
+                    }
+                });
 
-            return rc != 0
+                if (rc == 0 && !string.IsNullOrEmpty(resolvedSdk))
+                {
+                    SetEnvironmentVariableIfEmpty("DOTNET_HOST_PATH", dotnetPath);
+                    break;
+                }
+            }
+            
+            return string.IsNullOrEmpty(resolvedSdk)
                 ? throw new InvalidOperationException(SdkResolutionExceptionMessage(nameof(NativeMethods.hostfxr_resolve_sdk2)))
                 : resolvedSdk;
         }
 
-        private static string ResolveDotnetPath()
+        private static IList<string> ResolveDotnetPathCandidates()
         {
-            string? dotnetPath = GetDotnetPathFromROOT();
+            var pathCandidates = new List<string?>
+            {
+                GetDotnetPathFromROOT()
+            };
+
+            string? dotnetExePath = GetCurrentProcessPath();
+            bool isRunFromDotnetExecutable = !string.IsNullOrEmpty(dotnetExePath)
+                && Path.GetFileName(dotnetExePath).Equals(s_exeName, StringComparison.InvariantCultureIgnoreCase);
 
-            if (string.IsNullOrEmpty(dotnetPath))
+            if (isRunFromDotnetExecutable)
             {
-                string? dotnetExePath = GetCurrentProcessPath();
-                bool isRunFromDotnetExecutable = !string.IsNullOrEmpty(dotnetExePath)
-                    && Path.GetFileName(dotnetExePath).Equals(s_exeName, StringComparison.InvariantCultureIgnoreCase);
-
-                dotnetPath = isRunFromDotnetExecutable
-                    ? Path.GetDirectoryName(dotnetExePath)
-                    : FindDotnetPathFromEnvVariable("DOTNET_HOST_PATH")
-                        ?? FindDotnetPathFromEnvVariable("DOTNET_MSBUILD_SDK_RESOLVER_CLI_DIR")
-                        ?? GetDotnetPathFromPATH();
+                pathCandidates.Add(Path.GetDirectoryName(dotnetExePath));
             }
 
-            return string.IsNullOrEmpty(dotnetPath)
+            pathCandidates.Add(FindDotnetPathFromEnvVariable("DOTNET_HOST_PATH"));
+            pathCandidates.Add(FindDotnetPathFromEnvVariable("DOTNET_MSBUILD_SDK_RESOLVER_CLI_DIR"));
+            pathCandidates.Add(GetDotnetPathFromPATH());
+
+            IList<string> filteredPathCandidates = pathCandidates.Where(pk => !string.IsNullOrEmpty(pk))
+                .Select(s => s!)
+                .ToList();
```

Comment:
nit: You can do this filtering right when adding to the list for a bit of extra efficiency. Maybe with a local function like:

```C#
    // ...
    
    AddIfValid(GetDotnetPathFromROOT());
    // ...
    AddIfValid(FindDotnetPathFromEnvVariable("DOTNET_HOST_PATH"));
    AddIfValid(FindDotnetPathFromEnvVariable("DOTNET_MSBUILD_SDK_RESOLVER_CLI_DIR"));
    AddIfValid(GetDotnetPathFromPATH());

    // ...

    void AddIfValid(string? path)
    {
        if (!string.IsNullOrEmpty(path))
        {
            pathCandidates.Add(path);
        }
    }
```

Reply:
done